cdc_iterator.go

```diff
@@ -0,0 +1,293 @@
+	// the iterator teardown method is called
+	go func() {
+		<-iterator.canalTomb.Dead()
+		close(iterator.data)
+	}()
+
+	return iterator, nil
+}
+
+func (c *cdcIterator) getStartPosition(config cdcIteratorConfig) (mysql.Position, error) {
+	if config.position != nil {
+		pos, err := common.ParseSDKPosition(config.position)
+		if err != nil {
+			return mysql.Position{}, fmt.Errorf("failed to parse position: %w", err)
+		}
+		if pos.Kind == common.PositionTypeSnapshot {
+			return mysql.Position{}, fmt.Errorf("invalid position type: %s", pos.Kind)
+		}
+
+		return pos.CdcPosition.Position, nil
+	}
+
+	masterPos, err := c.canal.GetMasterPos()
+	if err != nil {
+		return mysql.Position{}, fmt.Errorf("failed to get master position: %w", err)
+	}
+
+	return masterPos, nil
+}
+
+func (c *cdcIterator) runCanal(ctx context.Context, startPos mysql.Position) error {
+	// buffered to allow the goroutine to exit when the iterator is torn down
+	errChan := make(chan error, 1)
+	go func() {
+		handler := &cdcEventHandler{
+			ctx:  ctx,
+			data: c.data,
+		}
+		c.canal.SetEventHandler(handler)
+		errChan <- c.canal.RunFrom(startPos)
+	}()
+
+	select {
+	case <-ctx.Done():
+		c.canal.Close()
+		return fmt.Errorf("context cancelled from runCanal: %w", ctx.Err())
+	case err := <-errChan:
+		return fmt.Errorf("failed to run canal: %w", err)
+	}
+}
+
+func (c *cdcIterator) Ack(context.Context, sdk.Position) error {
+	c.acks.Done()
+	return nil
+}
+
+func (c *cdcIterator) Next(ctx context.Context) (sdk.Record, error) {
+	select {
+	case <-ctx.Done():
+		return sdk.Record{}, fmt.Errorf("context cancelled: %w", ctx.Err())
+	case data, ok := <-c.data:
+		if !ok { // closed
+			if err := c.canalTomb.Err(); err != nil {
+				return sdk.Record{}, fmt.Errorf("canal exited unexpectedly: %w", err)
+			}
+			if err := c.acks.Wait(ctx); err != nil {
+				return sdk.Record{}, fmt.Errorf("failed to wait for acks: %w", err)
+			}
+			return sdk.Record{}, fmt.Errorf("cdc iterator teared down")
+		}
+
+		rec, err := c.buildRecord(data)
+		if err != nil {
+			return sdk.Record{}, fmt.Errorf("failed to build record: %w", err)
+		}
+
+		c.acks.Add(1)
+		return rec, nil
+	}
+}
+
+func (c *cdcIterator) Teardown(context.Context) error {
+	if c.canalTomb != nil {
+		c.canalTomb.Kill(errors.New("tearing down snapshot iterator"))
+	}
+
+	return nil
+}
+
+func buildPayload(columns []schema.TableColumn, rows []any) sdk.StructuredData {
+	payload := sdk.StructuredData{}
+	for i, col := range columns {
+		val := rows[i]
+		if s, ok := val.(string); ok {
+			// I don't know why exactly, but "github.com/go-mysql-org/go-mysql/canal"
+			// returns a string for timestamp columns without timezone. This is a hack
+			// to format the string back into an UTC string.
+			// TODO: investigate this further. Is this a bug in canal?
+			val = tryParseCanalStrDate(s)
+		}
+
+		payload[col.Name] = common.FormatValue(val)
+	}
+	return payload
+}
+
+func tryParseCanalStrDate(s string) string {
```

Comment:
Wait, so we parse the time and then format it? What's the purpose of this? 🧐

Reply:
Covering a configuration option that I missed. Just removed it.